sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,210 @@
+"""The AMQPMixin wraps RabbitMQ use into a request handler, with
+methods to speed the development of publishing RabbitMQ messages.
+
+"""
+from datetime import timedelta
+from pika.adapters.tornado_connection import TornadoConnection
+import logging
+import os
+import pika
+
+from tornado import gen, locks, web, ioloop
+
+LOGGER = logging.getLogger(__name__)
+
+WAIT_SECONDS = os.environ.get('WAIT_SECONDS', 1)
+WAIT_TIMEOUT = timedelta(seconds=int(WAIT_SECONDS))
+AMQP_STATES = {
+    'connected': 'CONNECTED',
+    'connecting': 'CONNECTING',
+    'disconnected': 'DISCONNECTED',
+}
+
+
+class AMQPMixin(web.RequestHandler):
+    """The request handler will connect to RabbitMQ on the first request,
+    blocking until the connection and channel are established. If RabbitMQ
+    closes it's connection to the app at any point, a connection attempt will
+    be made on the next request.
+
+    Expects the :envvar:`AMQP` environment variable to construct
+    :class:`pika.connection.URLParameters`.
+
+    https://github.com/gmr/tinman/blob/master/tinman/handlers/rabbitmq.py
+
+    """
+    amqp_channel_ready = locks.Event()
+    amqp_ready = locks.Event()
```

Comment:
Right but you have a connection chain of:

Open Connection -> on Open of Connection -> Open Channel -> on Open of Channel

So why not just have a single event that is set when your "on Open of Channel" occurs. You don't need to track the state of connected but with no open channels. When a channel is closed, you can clear the event and open the channel again.